modules/storages/app/common/storages/peripherals/storage_interaction/authentication_strategies/nextcloud_strategies.rb

```diff
@@ -37,10 +37,20 @@ module NextcloudStrategies
             ::Storages::Peripherals::StorageInteraction::AuthenticationStrategies::BasicAuth.strategy
           end
 
-          UserBound = ->(user:) do
-            ::Storages::Peripherals::StorageInteraction::AuthenticationStrategies::OAuthUserToken
-              .strategy
-              .with_user(user)
+          UserBound = ->(user:, storage:) do
+            sso_preferred = storage.audience.present? && user.authentication_provider.present?
+
+            if sso_preferred
+              ::Storages::Peripherals::StorageInteraction::AuthenticationStrategies::SsoUserToken
+                .strategy
+                .with_user(user)
+            elsif storage.oauth_configuration.present?
+              ::Storages::Peripherals::StorageInteraction::AuthenticationStrategies::OAuthUserToken
+                .strategy
+                .with_user(user)
+            else
+              ::Storages::Peripherals::StorageInteraction::AuthenticationStrategies::Noop.strategy
```

Comment:
:yellow_circle:  As discussed in the Draft PR, we should still agree on whether we want to keep using the `Noop` strategy here, which will eventually result in a request being made without any authentication and subsequently fail or whether we want to cause an error for this case.

My preference would be to fail loudly, since we do not expect to come into this situation and thus should fail early and explicitly.

Reply:
Yes, I remember. So, we have the noop strategy, which basically just yields the request block. This is needed for other places as well. This method here needs to return a strategy, and somehow it feels weird to raise an error here. So....

What about the idea of having one more strategy that MUST fail without doing a request? It is a common pattern. We could name it `AuthenticationStrategies::Error` or `AuthenticationStrategies::Failure`.

Reply:
Sounds good to me, naming wise I'd choose `Failure`.